src/passes/OptimizeInstructions.cpp

```diff
@@ -860,6 +819,86 @@ struct OptimizeInstructions
     return boolean;
   }
 
+  Expression* optimizeSelect(Select* curr) {
+    using namespace Match;
+    Builder builder(*getModule());
+    curr->condition = optimizeBoolean(curr->condition);
+    {
+      // Constant condition, we can just pick the correct side (barring side
+      // effects)
+      Expression *ifTrue, *ifFalse;
+      if (matches(curr, select(pure(&ifTrue), any(&ifFalse), i32(0)))) {
+        return ifFalse;
+      }
+      if (matches(curr, select(any(&ifTrue), any(&ifFalse), i32(0)))) {
+        return builder.makeSequence(builder.makeDrop(ifTrue), ifFalse);
+      }
+      if (matches(curr, select(any(&ifTrue), pure(&ifFalse), i32()))) {
+        // The condition must be non-zero
```

Comment:
wait, does `i32()` not just check for any `i32` but also a non-zero one?

Reply:
No, it's just non-zero here because zeroes would have matched above. I'll clarify that in the comment and also add an assert.